### ISSUE TITLE
Numbers as dynamic length strings in wire protocol

### ISSUE THREAD

**Issue by alippai** [vertica/vertica-python#398]
Is this the only option or is there a way to force a binary float32 / float64 / int64 type?

**Comment by sitingren** [comment]
@alippai I don't quite understand your question. Would you provide some examples?

**Comment by alippai** [comment]
Sure.
The floats are coming as string over the wire:
https://github.com/vertica/vertica-python/blob/master/vertica_python/vertica/column.py#L143

Is there a way to change the protocol, so fixed width data rows would be possible?

**Comment by sitingren** [comment]
The wire protocol is used for the delivery of what is stored in database to the client, it does not modify the data. So a float data comes as a string representation of its value and some metadata (e.g. datatype). 

So I think you have to modify your program to process the fetched data to fixed width, or change your data to be stored as fixed width type like `CHAR(n)`, or make a customization of vertica-python.

**Comment by alippai** [comment]
I'm pretty confident it's stored as float64 and int64 server-side and serialization into string (instead of bytes) happens.

https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/DataTypes/SQLDataTypes.htm
Floats supposed to be represented as "Signed 64-bit IEEE floating point number, requiring 8 bytes of storage", this is how it should be sent - as a 8 byte value. Why converting it to string and back? It's already fixed-width.

Let me know if there is a reference saying that numbers are stored as ASCII strings in Vertica.

>  or make a customization of vertica-python.

This was my intention, I'm happy to make the PR if there is a different protocol version (eg. for use with different drivers like ODBC or JDBC) which supports numbers as binary instead of string type.

Otherways I'm happy to wait for an inevitable Apache Arrow support :P

**Comment by tomwall** [comment]
The protocol support two operating modes -- string & binary -- but currently vertica-python only implements the string protocol. While postgres lets you change it on a per-column/type basis in the protocol, Vertica currently does not honor that and instead does it as a session-level setting. So unfortunately that means its an all-or-nothing implementation - all types must be implemented, both for reads (queries) and for writes (server-prepared statement bind values) in order for it to be viable. 

That makes it a reasonably large feature, but now that all the protocols for core functionality are implemented, it would be a good project to have on the roadmap. Converting binary format to & from python objects is a bit more complicated than strings, but it should have a good performance benefit, especially when you're dealing with lots of numeric data. 

To opt into that, you'd have to change the startup message (https://github.com/vertica/vertica-python/blob/master/vertica_python/vertica/messages/frontend_messages/startup.py#L73) to include 'binary_data_protocol' set to '1' (0 is the default, which is the string protocol), then implement a bunch of new serializers & deserializers. The binary encoding of a given type is the same as what's described in the NATIVE file format, just without the headers and other stuff (https://www.vertica.com/docs/10.0.x/HTML/Content/Authoring/AdministratorsGuide/BinaryFilesAppendix/ColumnDefinitions.htm).

**Comment by alippai** [comment]
Thanks, this is a perfect answer!

**Comment by alippai** [comment]
A quick experiment:
- started with 3 million rows, one float column, 36+ seconds to read it into a numpy array
- adding a `self.socket_as_file = socket.makefile('rb', 10*1024*1024)` and modifying `Connection.read_bytes(n)` to use it instantly adds 30-50% boost, it's down to 21-23seconds
- setting the `binary_data_protocol: '1'` saves 20% network bandwidth (51MB instead of 70MB on my sample), reading the result simply with `Connection.socket_as_file.read()` is faster too (0.7 seconds as binary vs 3-4 seconds as string)
- changing to `struct.unpack('>d', s)` from `float(s)` is not really noticeable, because parsing the whole message in python is simply slow, still 20+ seconds
- using `numpy.recarray(shape=(num_rows,), dtype=[header, size, field_count, field_size, value], buf)` on the raw buffer makes it superb, it's around 0.8s in _total_ (starting from `cursor.execute()` to the finished numpy array)

tl;dr it's possible to make it cca **50x faster**

The only drawback at this point is that I have to know the result set size (number of rows) to feed the numpy shape with it. @tomwall is there a way to query the result size before it arrives? (I don't see it in the rows description, but this might be an oversight)

**Comment by alonme** [comment]
is there any reason why this issue is closed?

sounds like a well needed possibility to boost performance
(@sitingren @tomwall)

**Comment by alonme** [comment]
hey @alippai, 
do you know if there are any plans to support one of these in this library?

**Comment by alippai** [comment]
No, I have zero information regarding any of this. I usually simply use the Turbodbc driver or the COPY protocol (also the built-in Parquet support is really nice) when fetching / pushing a lot of data and this lib when just moving data within Vertica. The Vertica ODBC driver is pretty good.

**Comment by alonme** [comment]
@alippai thanks a lot, will check that out!

**Comment by alonme** [comment]
@alippai 
what about this optimization?
it seems most of the time is spend in `read_bytes`, were your modifications only relevant to the binary protocol?

> adding a self.socket_as_file = socket.makefile('rb', 10*1024*1024) and modifying Connection.read_bytes(n) to use it instantly adds 30-50% boost, it's down to 21-23seconds

**Comment by sitingren** [comment]
@alonme There is a plan to implement and release `binary_data_protocol: '1'` within next few months.
